### PR TITLE
Parser for Mongolia

### PR DESCRIPTION
**Pull request by alixunderplatz**
- fetch production data
- fetch consumption data

### CODE REVIEW

**Comment by alixunderplatz** [comment]
I think the parser is ready for a check now, @VIKTORVAV99 :) haven't done this in a while - is there anything else I need to add except for the parser and capacity-stuff? In the past I think I'd have to specify the parser which is used for a zone somewhere...^^

I didn't add any warnings or exceptions for the moment. I guess we will see how well it works in reality :) datapoints of the data provider are lagging sometimes, but I haven't seen any gaps larger than 3 hours so far.

I calculated a carbon intensity for the unknown part consisting of coal, oil and hydro.

for 18:00 local time:
fetch_production() ->
{'zoneKey': 'MN', 'datetime': datetime.datetime(2023, 1, 31, 18, 0, tzinfo=tzfile('/usr/share/zoneinfo/Asia/Ulaanbaatar')), 'production': {'unknown': 1047.36, 'solar': 46.15, 'wind': 104.16}, 'storage': {}, 'source': 'https://ndc.energy.mn/'}
fetch_consumption() ->
{'zoneKey': 'MN', 'datetime': datetime.datetime(2023, 1, 31, 18, 0, tzinfo=tzfile('/usr/share/zoneinfo/Asia/Ulaanbaatar')), 'consumption': {'consumption': 1259.23}, 'storage': {}, 'source': 'https://ndc.energy.mn/'}

for 19:00 local time:
fetch_production() ->
{'zoneKey': 'MN', 'datetime': datetime.datetime(2023, 1, 31, 19, 0, tzinfo=tzfile('/usr/share/zoneinfo/Asia/Ulaanbaatar')), 'production': {'unknown': 1048.83, 'solar': 44.08, 'wind': 111.63}, 'storage': {}, 'source': 'https://ndc.energy.mn/'}
fetch_consumption() ->
{'zoneKey': 'MN', 'datetime': datetime.datetime(2023, 1, 31, 19, 0, tzinfo=tzfile('/usr/share/zoneinfo/Asia/Ulaanbaatar')), 'consumption': {'consumption': 1360.35}, 'storage': {}, 'source': 'https://ndc.energy.mn/'}

**Comment by VIKTORVAV99** [comment]
> I think the parser is ready for a check now, @VIKTORVAV99 :) haven't done this in a while - is there anything else I need to add except for the parser and capacity-stuff? In the past I think I'd have to specify the parser which is used for a zone somewhere...^^
> 
> I didn't add any warnings or exceptions for the moment. I guess we will see how well it works in reality :) datapoints of the data provider are lagging sometimes, but I haven't seen any gaps larger than 3 hours so far.
> 
> I calculated a carbon intensity for the unknown part consisting of coal, oil and hydro.
> 
> for 18:00 local time: fetch_production() -> {'zoneKey': 'MN', 'datetime': datetime.datetime(2023, 1, 31, 18, 0, tzinfo=tzfile('/usr/share/zoneinfo/Asia/Ulaanbaatar')), 'production': {'unknown': 1047.36, 'solar': 46.15, 'wind': 104.16}, 'storage': {}, 'source': '[https://ndc.energy.mn/'}](https://ndc.energy.mn/'%7D) fetch_consumption() -> {'zoneKey': 'MN', 'datetime': datetime.datetime(2023, 1, 31, 18, 0, tzinfo=tzfile('/usr/share/zoneinfo/Asia/Ulaanbaatar')), 'consumption': {'consumption': 1259.23}, 'storage': {}, 'source': '[https://ndc.energy.mn/'}](https://ndc.energy.mn/'%7D)
> 
> for 19:00 local time: fetch_production() -> {'zoneKey': 'MN', 'datetime': datetime.datetime(2023, 1, 31, 19, 0, tzinfo=tzfile('/usr/share/zoneinfo/Asia/Ulaanbaatar')), 'production': {'unknown': 1048.83, 'solar': 44.08, 'wind': 111.63}, 'storage': {}, 'source': '[https://ndc.energy.mn/'}](https://ndc.energy.mn/'%7D) fetch_consumption() -> {'zoneKey': 'MN', 'datetime': datetime.datetime(2023, 1, 31, 19, 0, tzinfo=tzfile('/usr/share/zoneinfo/Asia/Ulaanbaatar')), 'consumption': {'consumption': 1360.35}, 'storage': {}, 'source': '[https://ndc.energy.mn/'}](https://ndc.energy.mn/'%7D)

I'll try and do a full review as soon as possible! As for defining where the parser it used that is now done in the zone file (as you already have done) so no changes needed there! Logging and such is always a great addition to add if things don't work as expected (endpoint is unavailable for example) but it's not required, I'll take a closer look at this when I do the full review.

File: parsers/MN.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+def fetch_production(
+    zone_key: str = "MN",
+    session: Optional[Session] = None,
+    target_datetime: Optional[datetime] = None,
+    logger: Logger = getLogger(__name__),
+):
+    if target_datetime:
+        raise NotImplementedError("This parser is not yet able to parse past dates.")
+    with get(NDC_GENERATION) as response:
+        NDC_SOUP = BeautifulSoup(response.content, "html.parser")
```
Comment (VIKTORVAV99) [inline parsers/MN.py:25]: Should this use session.get() instead?
```suggestion
    with session.get(NDC_GENERATION) as response:
        NDC_SOUP = BeautifulSoup(response.content, "html.parser")
```

File: parsers/MN.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+from requests import Session, get
+
+#   MAIN_WEBSITE = https://ndc.energy.mn/
+NDC_GENERATION = "https://disnews.energy.mn/test/convert.php"
+TZ = "Asia/Ulaanbaatar"  # UTC+8
+
+
+def fetch_production(
+    zone_key: str = "MN",
+    session: Optional[Session] = None,
```
Comment (VIKTORVAV99) [inline parsers/MN.py:18]:
```suggestion
    session: Session = Session(),
```

File: parsers/MN.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        "storage": {},
+        "source": "https://ndc.energy.mn/",
+    }
+
+    return data
+
+
+def fetch_consumption(
+    zone_key: str = "MN",
+    session: Optional[Session] = None,
```
Comment (VIKTORVAV99) [inline parsers/MN.py:67]:
```suggestion
    session: Session = Session(),
```

File: parsers/MN.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    target_datetime: Optional[datetime] = None,
+    logger: Logger = getLogger(__name__),
+):
+    if target_datetime:
+        raise NotImplementedError("This parser is not yet able to parse past dates.")
+    with get(NDC_GENERATION) as response:
```
Comment (VIKTORVAV99) [inline parsers/MN.py:73]:
```suggestion
    with session.get(NDC_GENERATION) as response:
```

File: parsers/MN.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        NDC_SOUP = BeautifulSoup(response.content, "html.parser")
+
+    # Нийлбэр ачаалал
+    consumption_MW = float(NDC_SOUP.find_all("td")[6].text.strip(" МВт"))
+
+    dataset_consumption = {
+        'consumption': consumption_MW
+    }
```
Comment (VIKTORVAV99) [inline parsers/MN.py:81]:
```suggestion
```

File: parsers/MN.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    time_time = datetime.strptime(time_raw, '%H:%M:%S')
+    time_pretty = datetime.time(time_time)
+
+    time_combined = datetime.combine(date_pretty, time_pretty)
+    time_arrow = arrow.get(time_combined, tzinfo='Asia/Ulaanbaatar')
+
+    data = {
+        "zoneKey": zone_key,
+        "datetime": time_arrow.datetime,
+        "consumption": dataset_consumption,
```
Comment (VIKTORVAV99) [inline parsers/MN.py:95]:
```suggestion
        "consumption": consumption_mw,
```
This should be returned as a float (not a dict).

**Review by VIKTORVAV99 — COMMENTED** [review]
Most of the comment (except the last one) is nit picks and aren't strictly nessesary but would allign things with out other parsers.

It also looks like you need to run `poetry run format` to get the formatting right but byond that this parser is looking really good! Thanks for creating it!

**Comment by alixunderplatz** [comment]
Thanks @VIKTORVAV99 for your suggestions and fixes :-)
Should be fine now 🇲🇳 🔌 🗺️

fetch_production() ->
{'zoneKey': 'MN', 'datetime': datetime.datetime(2023, 2, 8, 12, 0, tzinfo=tzfile('/usr/share/zoneinfo/Asia/Ulaanbaatar')), 'production': {'unknown': 1110.51, 'solar': 26.19, 'wind': 29.07}, 'storage': {}, 'source': 'https://ndc.energy.mn/'}
fetch_consumption() ->
{'zoneKey': 'MN', 'datetime': datetime.datetime(2023, 2, 8, 12, 0, tzinfo=tzfile('/usr/share/zoneinfo/Asia/Ulaanbaatar')), 'consumption': 1289.9, 'storage': {}, 'source': 'https://ndc.energy.mn/'}

**Review by VIKTORVAV99 — APPROVED** [review]
Thanks again for making this parser!

I pushed some final cleanup changes, formatting and merged in master so this is good to go now!

**Comment by madsnedergaard** [comment]
So cool to get Mongolia on the map, thanks @alixunderplatz 😍 

Screenshot 2023-03-02 at 11 17 26

**Pull request merged**